src/engine/runtime.js

```diff
@@ -536,29 +521,52 @@ class Runtime extends EventEmitter {
-                        const menuItems = extensionInfo.menus[menuName];
-                        const convertedMenu = this._buildMenuForScratchBlocks(menuName, menuItems, categoryInfo);
-                        categoryInfo.menus.push(convertedMenu);
-                    }
-                }
-                for (const blockInfo of extensionInfo.blocks) {
-                    const convertedBlock = this._convertForScratchBlocks(blockInfo, categoryInfo);
-                    const opcode = convertedBlock.json.type;
-                    categoryInfo.blocks.push(convertedBlock);
-                    this._primitives[opcode] = convertedBlock.info.func;
-                    if (blockInfo.blockType === BlockType.HAT) {
-                        this._hats[opcode] = {edgeActivated: true}; /** @TODO let extension specify this */
-                    }
-                }
+                this._fillExtensionCategory(categoryInfo, extensionInfo);
                 extensionBlocks = extensionBlocks.concat(categoryInfo.blocks, categoryInfo.menus);
             }
         }
 
         this.emit(Runtime.BLOCKSINFO_UPDATE, extensionBlocks);
     }
 
+    /**
+     * Read extension information, convert menus and blocks, and store the results in the provided category object.
+     * @param {CategoryInfo} categoryInfo - the category to be filled
+     * @param {ExtensionMetadata} extensionInfo - the extension metadata to read
+     * @private
+     */
+    _fillExtensionCategory (categoryInfo, extensionInfo) {
+        for (const menuName in extensionInfo.menus) {
+            if (extensionInfo.menus.hasOwnProperty(menuName)) {
+                const menuItems = extensionInfo.menus[menuName];
+                const convertedMenu = this._buildMenuForScratchBlocks(menuName, menuItems, categoryInfo);
+                categoryInfo.menus.push(convertedMenu);
+            }
+        }
+        for (const blockInfo of extensionInfo.blocks) {
+            if (blockInfo === '---') {
+                categoryInfo.blocks.push(ConvertedSeparator);
+                continue;
+            }
+            try {
+                const convertedBlock = this._convertForScratchBlocks(blockInfo, categoryInfo);
+                const opcode = convertedBlock.json.type;
+                categoryInfo.blocks.push(convertedBlock);
+                if (blockInfo.blockType !== BlockType.EVENT) {
+                    this._primitives[opcode] = convertedBlock.info.func;
+                }
+                if (blockInfo.blockType === BlockType.EVENT || blockInfo.blockType === BlockType.HAT) {
+                    this._hats[opcode] = {
+                        edgeActivated: blockInfo.isEdgeActivated,
+                        restartExistingThreads: blockInfo.shouldRestartExistingThreads
+                    };
+                }
+            } catch (e) {
+                log.error('Error parsing block: ', {block: blockInfo, error: e});
+            }
+        }
+    }
+
```

Comment:
If we can avoid functions that modify state external to them when we can, I think it will be easier to understand how everything works in general. It is a little unavoidable in the blocks loop since we have to register the hats and block implementations, but could we at least break this up into two, one to get the menus and one to get the blocks?

Reply:
Thanks for catching this @rschamp. Good point. 👍

Reply:
@cwillisf This seems ok to do in follow-up. Do you mind filing an issue?

Reply:
Filed as #1043